<a href="https://colab.research.google.com/github/KzBrain/pythonKazbek/blob/master/HW_7_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
library(lubridate)
library(tidyverse)
library(survival)
install.packages('gtsummary')
library(gtsummary)
library(tidycmprsk)
devtools::install_github("zabore/condsurv") # author package
library(condsurv)


Attaching package: ‘lubridate’


The following objects are masked from ‘package:base’:

    date, intersect, setdiff, union


── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──
✔ dplyr   1.1.4     ✔ readr   2.1.5
✔ forcats 1.0.0     ✔ stringr 1.5.1
✔ ggplot2 3.4.4     ✔ tibble  3.2.1
✔ purrr   1.0.2     ✔ tidyr   1.3.1
── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ purrr::%||%()   masks base::%||%()
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()
ℹ Use the conflicted package (<http://conflicted.r-lib.org/>) to force all conflicts to become errors
Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)

#BlackLivesMatter


Attaching package: ‘tidycmprsk’


The following object is masked from ‘package:gtsummary’:

    trial





munsell    (0.5.0 -> 0.5.1) [CRAN]
farver     (2.1.1 -> 2.1.2) [CRAN]
isoband    (0.2.6 -> 0.2.7) [CRAN]
gtable     (0.3.4 -> 0.3.5) [CRAN]
tidyselect (1.2.0 -> 1.2.1) [CRAN]
ggplot2    (3.4.4 -> 3.5.1) [CRAN]


Installing 6 packages: munsell, farver, isoband, gtable, tidyselect, ggplot2

Installing packages into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)



── R CMD build ─────────────────────────────────────────────────────────────────
* checking for file ‘/tmp/RtmphTuI80/remotes29313359b97/zabore-condsurv-e618590/DESCRIPTION’ ... OK
* preparing ‘condsurv’:
* checking DESCRIPTION meta-information ... OK
* checking for LF line-endings in source and make files and shell scripts
* checking for empty or unneeded directories
Omitted ‘LazyData’ from DESCRIPTION
* building ‘condsurv_1.0.0.tar.gz’



Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)



In [ ]:
install.packages('tidycmprsk')

Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)

also installing the dependencies ‘Rcpp’, ‘V8’, ‘reactR’, ‘labelled’, ‘bigD’, ‘bitops’, ‘htmlwidgets’, ‘juicyjuice’, ‘markdown’, ‘reactable’, ‘broom.helpers’, ‘gt’, ‘cmprsk’, ‘gtsummary’, ‘hardhat’




In [ ]:
install.packages('condsurv')

Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)

Warning message:
“package ‘condsurv’ is not available for this version of R

A version of this package for your version of R might be available elsewhere,
see the ideas at
https://cran.r-project.org/doc/manuals/r-patched/R-admin.html#Installing-packages”
Warning message:
“Perhaps you meant ‘condSURV’ ?”


In [ ]:
colon_ds <- read_tsv('/content/sample_data/colon_dataset.tsv')

Rows: 1858 Columns: 16
── Column specification ────────────────────────────────────────────────────────
Delimiter: "\t"
chr  (1): rx
dbl (15): id, study, sex, age, obstruct, perfor, adhere, nodes, status, diff...

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.


In [ ]:
head(colon_ds)

id study rx      sex age obstruct perfor adhere nodes status differ extent
1 1  1     Lev+5FU 1   43  0        0      0      5     1      2      3     
2 1  1     Lev+5FU 1   43  0        0      0      5     1      2      3     
3 2  1     Lev+5FU 1   63  0        0      0      1     0      2      3     
4 2  1     Lev+5FU 1   63  0        0      0      1     0      2      3     
5 3  1     Obs     0   71  0        0      1      7     1      2      2     
6 3  1     Obs     0   71  0        0      1      7     1      2      2     
  surg node4 time etype
1 0    1     1521 2    
2 0    1      968 1    
3 0    0     3087 2    
4 0    0     3087 1    
5 0    1      963 2    
6 0    1      542 1

In [ ]:
colon_df <- colon_ds %>% select(time,status)
colon_df

time status
1    1521 1     
2     968 1     
3    3087 0     
4    3087 0     
5     963 1     
6     542 1     
7     293 1     
8     245 1     
9     659 1     
10    523 1     
11   1767 1     
12    904 1     
13    420 1     
14    229 1     
15   3192 0     
16   3192 0     
17   3173 0     
18   3173 0     
19   3308 0     
20   3308 0     
21   2908 0     
22   2908 0     
23   3309 0     
24   3309 0     
25   2085 1     
26   1130 1     
27   2910 1     
28   2231 1     
29   2754 0     
30   2754 0     
⋮    ⋮    ⋮     
1829 2114 0     
1830 2114 0     
1831 1262 1     
1832  438 1     
1833  259 1     
1834  122 1     
1835  323 1     
1836   78 1     
1837  678 1     
1838  532 1     
1839 1827 0     
1840 1827 0     
1841 1399 1     
1842  476 1     
1843 1022 1     
1844  329 1     
1845 2240 0     
1846 2240 0     
1847 2158 0     
1848 2158 0     
1849 1875 0     
1850 1875 0     
1851 2154 0     
1852 2154 0     
1853 1018 1     
1854  851 1     
1855 2072 0     
1856 2072 0     
1857 1820 0     
1858 1820 0

In [ ]:
colon_df %>%
  tbl_summary(statistic = status ~ '{n}/{N}') %>%
    add_stat_label() %>%
    bold_labels()

<div id="bkbyuyrepw" style="padding-left:0px;padding-right:0px;padding-top:10px;padding-bottom:10px;overflow-x:auto;overflow-y:auto;width:auto;height:auto;">
  <style>#bkbyuyrepw table {
  font-family: system-ui, 'Segoe UI', Roboto, Helvetica, Arial, sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji', 'Segoe UI Symbol', 'Noto Color Emoji';
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

#bkbyuyrepw thead, #bkbyuyrepw tbody, #bkbyuyrepw tfoot, #bkbyuyrepw tr, #bkbyuyrepw td, #bkbyuyrepw th {
  border-style: none;
}

#bkbyuyrepw p {
  margin: 0;
  padding: 0;
}

#bkbyuyrepw .gt_table {
  display: table;
  border-collapse: collapse;
  line-height: normal;
  margin-left: auto;
  margin-right: auto;
  color: #333333;
  font-size: 16px;
  font-weight: normal;
  font-style: normal;
  background-color: #FFFFFF;
  width: auto;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #A8A8A8;
  border-right-style: none;
  border-right-width: 2px;
  b

In [ ]:
Surv(colon_ds$time,colon_ds$status)[1:50]

 [1] 1521   968  3087+ 3087+  963   542   293   245   659   523  1767   904 
[13]  420   229  3192+ 3192+ 3173+ 3173+ 3308+ 3308+ 2908+ 2908+ 3309+ 3309+
[25] 2085  1130  2910  2231  2754+ 2754+ 3214+ 1323   406   258   522   389 
[37]  887   604  3329+ 3329+ 2789  2789+  739   527   709   348  2969+ 2969+
[49] 2889+ 2889+

In [ ]:
ds_1 <- survfit(Surv(time,status) ~ 1,data = colon_ds)
str(ds_1)

List of 16
 $ n        : int 1858
 $ time     : num [1:1067] 8 9 19 20 23 24 28 34 35 36 ...
 $ n.risk   : num [1:1067] 1858 1857 1856 1855 1854 ...
 $ n.event  : num [1:1067] 1 1 1 1 1 1 1 1 1 1 ...
 $ n.censor : num [1:1067] 0 0 0 0 1 1 0 0 0 0 ...
 $ surv     : num [1:1067] 0.999 0.999 0.998 0.998 0.997 ...
 $ std.err  : num [1:1067] 0.000538 0.000762 0.000933 0.001078 0.001205 ...
 $ cumhaz   : num [1:1067] 0.000538 0.001077 0.001616 0.002155 0.002694 ...
 $ std.chaz : num [1:1067] 0.000538 0.000761 0.000933 0.001077 0.001205 ...
 $ type     : chr "right"
 $ logse    : logi TRUE
 $ conf.int : num 0.95
 $ conf.type: chr "log"
 $ lower    : num [1:1067] 0.998 0.997 0.997 0.996 0.995 ...
 $ upper    : num [1:1067] 1 1 1 1 1 ...
 $ call     : language survfit(formula = Surv(time, status) ~ 1, data = colon_ds)
 - attr(*, "class")= chr "survfit"


In [ ]:
ds_1

Call: survfit(formula = Surv(time, status) ~ 1, data = colon_ds)

        n events median 0.95LCL 0.95UCL
[1,] 1858    920   2351    2018    2910

In [ ]:
ds_1 %>%
  ggsurvfit() +
  labs(
    x = 'days',y = 'probability'
  ) +
    add_confidence_interval() +
    add_risktable()

ERROR: Error in ggsurvfit(.): could not find function "ggsurvfit"


In [ ]:
summary(ds_1,times = 425)

Call: survfit(formula = Surv(time, status) ~ 1, data = colon_ds)

 time n.risk n.event survival std.err lower 95% CI upper 95% CI
  425   1502     347    0.813 0.00906        0.795        0.831

In [ ]:
ds_1 %>%
  tbl_survfit(times = 365.25,label = 'First year survival')

<div id="azycvercmh" style="padding-left:0px;padding-right:0px;padding-top:10px;padding-bottom:10px;overflow-x:auto;overflow-y:auto;width:auto;height:auto;">
  <style>#azycvercmh table {
  font-family: system-ui, 'Segoe UI', Roboto, Helvetica, Arial, sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji', 'Segoe UI Symbol', 'Noto Color Emoji';
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

#azycvercmh thead, #azycvercmh tbody, #azycvercmh tfoot, #azycvercmh tr, #azycvercmh td, #azycvercmh th {
  border-style: none;
}

#azycvercmh p {
  margin: 0;
  padding: 0;
}

#azycvercmh .gt_table {
  display: table;
  border-collapse: collapse;
  line-height: normal;
  margin-left: auto;
  margin-right: auto;
  color: #333333;
  font-size: 16px;
  font-weight: normal;
  font-style: normal;
  background-color: #FFFFFF;
  width: auto;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #A8A8A8;
  border-right-style: none;
  border-right-width: 2px;
  b

In [ ]:
ds_1 %>%
  tbl_survfit(times = 365.25*2,label = 'Second year survival')

<div id="qelysgzbfw" style="padding-left:0px;padding-right:0px;padding-top:10px;padding-bottom:10px;overflow-x:auto;overflow-y:auto;width:auto;height:auto;">
  <style>#qelysgzbfw table {
  font-family: system-ui, 'Segoe UI', Roboto, Helvetica, Arial, sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji', 'Segoe UI Symbol', 'Noto Color Emoji';
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

#qelysgzbfw thead, #qelysgzbfw tbody, #qelysgzbfw tfoot, #qelysgzbfw tr, #qelysgzbfw td, #qelysgzbfw th {
  border-style: none;
}

#qelysgzbfw p {
  margin: 0;
  padding: 0;
}

#qelysgzbfw .gt_table {
  display: table;
  border-collapse: collapse;
  line-height: normal;
  margin-left: auto;
  margin-right: auto;
  color: #333333;
  font-size: 16px;
  font-weight: normal;
  font-style: normal;
  background-color: #FFFFFF;
  width: auto;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #A8A8A8;
  border-right-style: none;
  border-right-width: 2px;
  b

In [ ]:
ds_1 %>%
  tbl_survfit(times = 365.25*3,label = 'Third year survival')

<div id="lvxxfakaqx" style="padding-left:0px;padding-right:0px;padding-top:10px;padding-bottom:10px;overflow-x:auto;overflow-y:auto;width:auto;height:auto;">
  <style>#lvxxfakaqx table {
  font-family: system-ui, 'Segoe UI', Roboto, Helvetica, Arial, sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji', 'Segoe UI Symbol', 'Noto Color Emoji';
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

#lvxxfakaqx thead, #lvxxfakaqx tbody, #lvxxfakaqx tfoot, #lvxxfakaqx tr, #lvxxfakaqx td, #lvxxfakaqx th {
  border-style: none;
}

#lvxxfakaqx p {
  margin: 0;
  padding: 0;
}

#lvxxfakaqx .gt_table {
  display: table;
  border-collapse: collapse;
  line-height: normal;
  margin-left: auto;
  margin-right: auto;
  color: #333333;
  font-size: 16px;
  font-weight: normal;
  font-style: normal;
  background-color: #FFFFFF;
  width: auto;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #A8A8A8;
  border-right-style: none;
  border-right-width: 2px;
  b

In [ ]:
colon_ds %>%
  filter(status == 1) %>%
  summarize(median_surv = median(time))

median_surv
1 577.5

In [ ]:
survdiff(Surv(time,status) ~ rx,data = colon_ds)

Call:
survdiff(formula = Surv(time, status) ~ rx, data = colon_ds)

             N Observed Expected (O-E)^2/E (O-E)^2/V
rx=Lev     620      333      295      4.93      7.26
rx=Lev+5FU 608      242      326     21.61     33.54
rx=Obs     630      345      299      7.01     10.40

 Chisq= 33.6  on 2 degrees of freedom, p= 5e-08 

Есть статистический значимые различия выживаемости в зависимости от применение лекарств о чем свидетельствует маленкое значение p= 5e-08

In [ ]:
# Отфильтровка данных для группы лечения 'Lev'
lev_group <- subset(colon_ds, rx == "Lev")

In [ ]:
survdiff(Surv(time,status) ~ sex,data = lev_group)

Call:
survdiff(formula = Surv(time, status) ~ sex, data = lev_group)

        N Observed Expected (O-E)^2/E (O-E)^2/V
sex=0 266      137      147     0.619      1.11
sex=1 354      196      186     0.486      1.11

 Chisq= 1.1  on 1 degrees of freedom, p= 0.3 

In [ ]:
# Отфильтровка данных для группы лечения 'Lev+5FU'
lev5fu_group <- subset(colon_ds, rx == "Lev+5FU")

In [ ]:
survdiff(Surv(time,status) ~ sex,data = lev5fu_group)

Call:
survdiff(formula = Surv(time, status) ~ sex, data = lev5fu_group)

        N Observed Expected (O-E)^2/E (O-E)^2/V
sex=0 326      149      122      5.88      11.9
sex=1 282       93      120      6.00      11.9

 Chisq= 11.9  on 1 degrees of freedom, p= 6e-04 

Есть статистический значимые различия выживаемости между полами которые используют одну и ту же терапию p= 6e-04

In [ ]:
# Отфильтровка данных для группы лечения 'Obs'
obs_group <- subset(colon_ds, rx == "Obs")

In [ ]:
survdiff(Surv(time,status) ~ sex,data = obs_group)

Call:
survdiff(formula = Surv(time, status) ~ sex, data = obs_group)

        N Observed Expected (O-E)^2/E (O-E)^2/V
sex=0 298      158      164     0.203     0.387
sex=1 332      187      181     0.183     0.387

 Chisq= 0.4  on 1 degrees of freedom, p= 0.5 

In [ ]:
fit <- survdiff(Surv(time,status) ~ rx,data = colon_ds)

In [ ]:
install.packages('survminer')

Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)

also installing the dependencies ‘numDeriv’, ‘SparseM’, ‘MatrixModels’, ‘minqa’, ‘nloptr’, ‘RcppEigen’, ‘carData’, ‘abind’, ‘pbkrtest’, ‘quantreg’, ‘lme4’, ‘corrplot’, ‘car’, ‘png’, ‘jpeg’, ‘ggrepel’, ‘ggsci’, ‘cowplot’, ‘ggsignif’, ‘polynom’, ‘rstatix’, ‘exactRankTests’, ‘mvtnorm’, ‘KMsurv’, ‘zoo’, ‘km.ci’, ‘xtable’, ‘gridtext’, ‘ggpubr’, ‘gridExtra’, ‘maxstat’, ‘survMisc’, ‘ggtext’




In [ ]:
library(survminer)

Loading required package: ggpubr


Attaching package: ‘survminer’


The following object is masked from ‘package:survival’:

    myeloma




In [ ]:
install.packages('ggsurv')

Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)

Warning message:
“package ‘ggsurv’ is not available for this version of R

A version of this package for your version of R might be available elsewhere,
see the ideas at
https://cran.r-project.org/doc/manuals/r-patched/R-admin.html#Installing-packages”


In [ ]:
survminer::ggsurvplot(fit,pval = TRUE,pval.method = TRUE)

ERROR: Error in survminer::ggsurvplot(fit, pval = TRUE, pval.method = TRUE): object 'ggsurv' not found


In [ ]:
head(colon_ds)

id study rx      sex age obstruct perfor adhere nodes status differ extent
1 1  1     Lev+5FU 1   43  0        0      0      5     1      2      3     
2 1  1     Lev+5FU 1   43  0        0      0      5     1      2      3     
3 2  1     Lev+5FU 1   63  0        0      0      1     0      2      3     
4 2  1     Lev+5FU 1   63  0        0      0      1     0      2      3     
5 3  1     Obs     0   71  0        0      1      7     1      2      2     
6 3  1     Obs     0   71  0        0      1      7     1      2      2     
  surg node4 time etype
1 0    1     1521 2    
2 0    1      968 1    
3 0    0     3087 2    
4 0    0     3087 1    
5 0    1      963 2    
6 0    1      542 1

In [ ]:
cox_model <- coxph(Surv(time,status)~age + rx, data = colon_ds)
summary(cox_model)

Call:
coxph(formula = Surv(time, status) ~ age + rx, data = colon_ds)

  n= 1858, number of events= 920 

               coef exp(coef)  se(coef)      z Pr(>|z|)    
age       -0.002057  0.997946  0.002807 -0.733    0.464    
rxLev+5FU -0.419327  0.657489  0.084520 -4.961    7e-07 ***
rxObs      0.020050  1.020253  0.076837  0.261    0.794    
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

          exp(coef) exp(-coef) lower .95 upper .95
age          0.9979     1.0021    0.9925    1.0035
rxLev+5FU    0.6575     1.5209    0.5571    0.7759
rxObs        1.0203     0.9801    0.8776    1.1861

Concordance= 0.544  (se = 0.01 )
Likelihood ratio test= 35.77  on 3 df,   p=8e-08
Wald test            = 33.65  on 3 df,   p=2e-07
Score (logrank) test = 34.16  on 3 df,   p=2e-07


Лечение Lev+5FU значимо снижает риск смерти (HR = 0.657, p = 7e-07)

In [ ]:
cox_model2 <- coxph(Surv(time,status)~surg + rx + node4, data = colon_ds)
summary(cox_model2)

Call:
coxph(formula = Surv(time, status) ~ surg + rx + node4, data = colon_ds)

  n= 1858, number of events= 920 

              coef exp(coef) se(coef)      z Pr(>|z|)    
surg       0.24449   1.27697  0.07214  3.389 0.000701 ***
rxLev+5FU -0.41204   0.66230  0.08457 -4.872 1.11e-06 ***
rxObs      0.01316   1.01325  0.07696  0.171 0.864229    
node4      0.91065   2.48595  0.06797 13.398  < 2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

          exp(coef) exp(-coef) lower .95 upper .95
surg         1.2770     0.7831    1.1086    1.4709
rxLev+5FU    0.6623     1.5099    0.5611    0.7817
rxObs        1.0132     0.9869    0.8714    1.1782
node4        2.4859     0.4023    2.1759    2.8402

Concordance= 0.635  (se = 0.009 )
Likelihood ratio test= 209.2  on 4 df,   p=<2e-16
Wald test            = 223.7  on 4 df,   p=<2e-16
Score (logrank) test = 236.4  on 4 df,   p=<2e-16


Операция (surg) значимо увеличивает риск смерти (HR = 1.277, p = 0.000701).
Лечение Lev+5FU (rxLev+5FU) значимо снижает риск смерти (HR = 0.662, p = 1.11e-06).
Отсутствие лечения (rxObs) не оказывает значимого влияния на выживаемость (HR = 1.013, p = 0.864).
Наличие метастазов в 4 или более лимфоузлах (node4) значимо увеличивает риск смерти (HR = 2.486, p < 2e-16)